integrations/tensorflow/bindings/python/pyiree/tf/support/tf_test_utils.py

```diff
@@ -429,12 +440,259 @@ def get_target_backends():
+
+  @staticmethod
+  def _check_same(ref, tar, rtol, atol):
+    """Checks that ref and tar have identical datastructures and values."""
+    # Check for matching types.
+    if not isinstance(tar, type(ref)):
+      logging.error(
+          "Expected ref and tar to have the same type but got '%s' and '%s'",
+          type(ref), type(tar))
+      return False
+
+    if ref is None:
+      # Nothing to compare (e.g. the called method had no outputs).
+      return True
+
+    # Recursive check for dicts.
+    if isinstance(ref, dict):
+      if ref.keys() != tar.keys():
+        logging.error(
+            "Expected ref and tar to have the same keys, but got '%s' and '%s'",
+            ref.keys(), tar.keys())
+        return False
+      # Check that all of the dictionaries' values are the same.
+      for key in ref:
+        if not TracedModule._check_same(ref[key], tar[key], rtol, atol):
+          return False
+
+    # Recursive check for iterables.
+    elif isinstance(ref, list) or isinstance(ref, tuple):
+      if len(ref) != len(tar):
+        logging.error(
+            "Expected ref and tar to have the same length, but got %s and %s",
+            len(ref), len(tar))
+        return False
+      # Check that all of the iterables' values are the same.
+      for i in range(len(ref)):
+        if not TracedModule._check_same(ref[i], tar[i], rtol, atol):
+          return False
+
+    # Base check for numpy arrays.
+    elif isinstance(ref, np.ndarray):
+      if isinstance(ref.flat[0], np.floating):
+        return np.allclose(ref, tar, rtol=rtol, atol=atol)
+      else:
+        return np.array_equal(ref, tar)
+
+    # If outputs end up here then an extra branch for that type should be added.
+    else:
+      raise TypeError(f"Encountered results with unexpected type {type(ref)}")
+    return True
+
+  def _get_trace_dir(self, artifacts_dir):
+    trace_dir = os.path.join(artifacts_dir, "traces", self.trace_name)
+    if not os.path.exists(trace_dir):
+      os.makedirs(trace_dir)
+    return trace_dir
+
+  def save_plaintext(self, artifacts_dir, summarize=True):
+    """Saves a human-readable string representation of this trace to disk.
+
+    Args:
+      artifacts_dir: the base directory to save the trace in.
+      summarize: a bool controlling whether numpy should summarize the inputs
+        and outputs if they're large. Setting this to False is very slow for
+        large outputs.
+    """
+    prior_printoptions = np.get_printoptions()
+    np.set_printoptions(
+        linewidth=NUMPY_LINEWIDTH,
+        threshold=None if summarize else sys.maxsize,
+        edgeitems=10)  # Can show more items since they won't clutter the logs.
+
+    trace_dir = self._get_trace_dir(artifacts_dir)
+    path = os.path.join(trace_dir, f"plaintext__{self.backend}.txt")
+    with open(path, "w") as f:
+      f.write(str(self))
+      f.write("\n")
+
+    np.set_printoptions(**prior_printoptions)
+
+  def save_pickle(self, artifacts_dir):
```

Comment:
This will be removed once we have a proper serialization format? Maybe add a TODO?

(Splitting out Trace will simplify designing the layering here)

Reply:
Sounds good! We'll want a way to load that at the python level anyway. Maybe via a `staticmethod` like `Trace.load(path)`.

Reply:
I just removed it for the time being.